### ISSUE TITLE
General feedback

### ISSUE THREAD

**Issue by inteon** [pantoniou/go-fyaml#1]
I think it's great that [libfyaml](https://github.com/pantoniou/libfyaml) adheres so well to the YAML 1.2 spec.
I fear however, that the complexity and possible lack of safety that comes from interfacing with a c library in go would be a reason for a lot of developers to not choose this library over the current most popular library `gopkg.in/yaml` library.
Please note that I'm not an expert in this field, so feel free to correct me:
Golang does not use libc; libfyaml does use libc -> this adds unnecessary code overhead + might make it less portable to new platforms. It seems like cgo does not compile easily to Windows.

I think `gopkg.in/yaml`'s strong-points are:
- is proven to work well
- has very limited API, that is stable!

What is still missing however:
- some important config options are missing (like list indentation settings)
- errors are too basic (missing location in document/ no difference between fatal and non-fatal errors)
- spec is not adhered to 100%
- some of its internal components are not exposed via the API (see [carvel-ytt](https://github.com/vmware-tanzu/carvel-ytt/tree/develop/pkg/yamlmeta/internal/yaml.v2))

https://github.com/goccy/go-yaml is a library that fixes some of these pain-points, I'm however not sure how well it implements the YAML spec.

In the future, the go community probably will require a solution for the fact that `gopkg.in/yaml` is becoming unmaintained, so it's quite relevant to start looking for the best alternatives/ solutions already.

**Comment by pantoniou** [comment]
> I think it's great that [libfyaml](https://github.com/pantoniou/libfyaml) adheres so well to the YAML 1.2 spec. I fear however, that the complexity and possible lack of safety that comes from interfacing with a c library in go would be a reason for a lot of developers to not chose this library over the current most popular library `gopkg.in/yaml` library. Please note that I'm not an expert in this field, so feel free to correct me: Golang does not use libc; libfyaml does use libc -> this adds unnecessary code overhead + might make it less portable to new platforms. It seems like cgo does not compile easily to Windows.
>

I agree that some degree of 'safety' of whatever you get to call is compromised when using cgo in golang projects.
However whether it is significant enough for every project it is doubtful. Many projects use cgo modules as a matter a fact, so for them this problem is moot.
 
Regarding the complexity, this is not indented as a 'port' of libfyaml to go; the user facing API is exactly the same (or possibly better) than the currently favored yaml implementation.

For demonstration purposes this is the canonical sample app for the WIP on this.

https://github.com/pantoniou/go-fyaml-sample

Notice there is no need to CGO, and the API is exactly the same.

> I think `gopkg.in/yaml`'s strong-points are:
> 
>     * is proven to work well
> 
>     * has very limited API, that is stable!
> 
> 

Taking into account how badly it does handling the test-suite I could argue against the first point.
I can grant #2.

> What is still missing however:
> 
>     * some important config options are missing (like list indentation settings)
> 
>     * errors are too basic (missing location in document/ no difference between fatal and non-fatal errors)
> 
>     * spec is not adhered to 100%
> 
>     * some of its internal components are not exposed via the API (see [carvel-ytt](https://github.com/vmware-tanzu/carvel-ytt/tree/develop/pkg/yamlmeta/internal/yaml.v2))
> 
> 

From what I can tell, the internals are just an old version of libyaml that has been converted to go, and beaten enough to compile.

> https://github.com/goccy/go-yaml is a library that fixes some of these pain-points, I'm however not sure how well it implements the YAML spec.
>

Excellent, I will take a look, and I will try to make a test-suite harness for it. 

The problem I see with both of these YAML implementations is that there is no effort done to adhere to the spec.
And believe me when I say that even getting to a point where the errors are merely corner cases will take a long of time.

I think this is a shame to have the go community stuck without a good solution for too long.

> In the future, the go community probably will require a solution for the fact that `gopkg.in/yaml` is becoming unmaintained, so it's quite relevant to start looking for the best alternatives/ solutions already.

And finally this tidbit is the most disappointing.

We have pretty extensive plans and feature for the next versions of YAML and I think having golang as a first class spec implementation is pretty important to me.

So watch this space, and I do thank you for taking the time to respond with your feedback.
